File: server/src/main/kotlin/com/coooolfan/unirhy/controller/TaskController.kt
package com.coooolfan.unirhy.controller

import cn.dev33.satoken.annotation.SaCheckLogin
import cn.dev33.satoken.annotation.SaCheckRole
import com.coooolfan.unirhy.config.ROLE_ADMIN
import com.coooolfan.unirhy.error.CommonException
import com.coooolfan.unirhy.error.TaskException
import com.coooolfan.unirhy.model.AsyncTask
import com.coooolfan.unirhy.model.by
import com.coooolfan.unirhy.service.task.AsyncTaskService
import com.coooolfan.unirhy.service.task.TaskStatusCounts
import com.coooolfan.unirhy.service.task.common.TaskStatus
import jakarta.servlet.http.HttpServletResponse
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import tools.jackson.databind.JsonNode

data class TaskCreateRequest(
    val namespace: String,
    val taskType: String,
    val payload: JsonNode,
)

data class TaskStatusPatchRequest(val status: TaskStatus)

data class TaskStatusTransitionRequest(
    val namespace: String,
    val taskType: String,
    val sourceStatuses: Set<TaskStatus>,
    val targetStatus: TaskStatus,
)

data class TaskStatusTransitionResponse(val transitioned: Int)

data class TaskDetailResponse(
    val task: @FetchBy("DEFAULT_TASK_FETCHER", ownerType = TaskController::class) AsyncTask,
    val childTaskCounts: TaskStatusCounts,
)

/** 统一任务资源管理接口。 */
@SaCheckLogin
@RestController
@RequestMapping("/api/tasks")
class TaskController(
    private val taskService: AsyncTaskService,
) {
    /** 创建一个根任务。 */
    @PostMapping
    @SaCheckRole(ROLE_ADMIN)
    @ResponseStatus(HttpStatus.ACCEPTED)
    @Throws(
        CommonException.Forbidden::class,
        TaskException.InvalidTaskKey::class,
        TaskException.InvalidParams::class,
        TaskException.DefinitionNotFound::class,
        TaskException.PluginUnavailable::class,
    )
    fun createTask(
        @RequestBody request: TaskCreateRequest,
        response: HttpServletResponse,
    ): @FetchBy("DEFAULT_TASK_FETCHER") AsyncTask {
        val task = taskService.create(
            request.namespace,
            request.taskType,
            request.payload,
            DEFAULT_TASK_FETCHER,
        )
        response.setHeader("Location", "/api/tasks/${task.id}")
        return task
    }

    @GetMapping
    fun listTasks(
        @RequestParam(required = false) parentId: Long?,
        @RequestParam(required = false, defaultValue = "false") rootsOnly: Boolean,
        @RequestParam(required = false) namespace: String?,
        @RequestParam(required = false) taskType: String?,
        @RequestParam(required = false) statuses: List<TaskStatus>?,
        @RequestParam(required = false) pageIndex: Int?,
        @RequestParam(required = false) pageSize: Int?,
    ): Page<@FetchBy("DEFAULT_TASK_FETCHER") AsyncTask> =
        taskService.list(
            parentId = parentId,
            rootsOnly = rootsOnly,
            namespace = namespace,
            taskType = taskType,
            statuses = statuses ?: emptyList(),
            pageIndex = pageIndex ?: 0,
            pageSize = pageSize ?: 20,
            fetcher = DEFAULT_TASK_FETCHER,
        )

    @GetMapping("/{id}")
    @Throws(TaskException.TaskNotFound::class)
    fun getTask(@PathVariable id: Long): TaskDetailResponse = TaskDetailResponse(
        task = taskService.get(id, DEFAULT_TASK_FETCHER),
        childTaskCounts = TaskStatusCounts.from(taskService.childStatusCounts(id)),
    )

    @GetMapping("/{id}/tree")
    @Throws(TaskException.TaskNotFound::class)
    fun getTaskTree(@PathVariable id: Long): @FetchBy("TASK_TREE_FETCHER") AsyncTask =
        taskService.get(id, TASK_TREE_FETCHER)

    @PostMapping("/status-transitions")
    @SaCheckRole(ROLE_ADMIN)
    @Throws(
        CommonException.Forbidden::class,
        TaskException.InvalidTaskKey::class,
        TaskException.PluginUnavailable::class,
        TaskException.StatusConflict::class,
    )
    fun transitionTaskStatuses(
        @RequestBody request: TaskStatusTransitionRequest,
    ): TaskStatusTransitionResponse = TaskStatusTransitionResponse(
        taskService.transitionStatuses(
            namespace = request.namespace,
            taskType = request.taskType,
            sourceStatuses = request.sourceStatuses,
            targetStatus = request.targetStatus,
        ),
    )

    /** 仅接受 [TaskStatus.PENDING]（重排队）与 [TaskStatus.CANCELLED]（取消），其余返回 `TASK:STATUS_CONFLICT` */
    @PatchMapping("/{id}")
    @SaCheckRole(ROLE_ADMIN)
    @Throws(
        CommonException.Forbidden::class,
        TaskException.TaskNotFound::class,
        TaskException.PluginUnavailable::class,
        TaskException.StatusConflict::class,
    )
    fun patchTask(
        @PathVariable id: Long,
        @RequestBody request: TaskStatusPatchRequest,
    ): @FetchBy("DEFAULT_TASK_FETCHER") AsyncTask =
        taskService.patchStatus(id, request.status, DEFAULT_TASK_FETCHER)

    companion object {
        val DEFAULT_TASK_FETCHER = newFetcher(AsyncTask::class).by {
            allScalarFields()
            parentId()
        }

        val TASK_TREE_FETCHER = newFetcher(AsyncTask::class).by {
            allScalarFields()
            parentId()
            `childTasks*`()
        }
    }
}

File: web/src/__generated/services/TaskController.ts
import type {Executor} from '../';
import type {AsyncTaskDto} from '../model/dto/';
import type {TaskStatus} from '../model/enums/';
import type {
    Page, 
    TaskCreateRequest, 
    TaskDetailResponse, 
    TaskStatusPatchRequest, 
    TaskStatusTransitionRequest, 
    TaskStatusTransitionResponse
} from '../model/static/';

/**
 * 统一任务资源管理接口。
 */
export class TaskController {
    
    constructor(private executor: Executor) {}
    
    /**
     * 创建一个根任务。
     */
    readonly createTask: (options: TaskControllerOptions['createTask']) => Promise<
        AsyncTaskDto['TaskController/DEFAULT_TASK_FETCHER']
    > = async(options) => {
        let _uri = '/api/tasks';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<AsyncTaskDto['TaskController/DEFAULT_TASK_FETCHER']>;
    }
    
    readonly getTask: (options: TaskControllerOptions['getTask']) => Promise<
        TaskDetailResponse
    > = async(options) => {
        let _uri = '/api/tasks/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<TaskDetailResponse>;
    }
    
    readonly getTaskTree: (options: TaskControllerOptions['getTaskTree']) => Promise<
        AsyncTaskDto['TaskController/TASK_TREE_FETCHER']
    > = async(options) => {
        let _uri = '/api/tasks/';
        _uri += encodeURIComponent(options.id);
        _uri += '/tree';
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<AsyncTaskDto['TaskController/TASK_TREE_FETCHER']>;
    }
    
    readonly listTasks: (options: TaskControllerOptions['listTasks']) => Promise<
        Page<AsyncTaskDto['TaskController/DEFAULT_TASK_FETCHER']>
    > = async(options) => {
        let _uri = '/api/tasks';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.parentId;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'parentId='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.rootsOnly;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'rootsOnly='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.namespace;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'namespace='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.taskType;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'taskType='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.statuses;
        if (_value !== undefined && _value !== null) {
            for (const _item of _value) {
                _uri += _separator
                _uri += 'statuses='
                _uri += encodeURIComponent(_item);
                _separator = '&';
            }
        }
        _value = options.pageIndex;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'pageIndex='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.pageSize;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'pageSize='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<Page<AsyncTaskDto['TaskController/DEFAULT_TASK_FETCHER']>>;
    }
    
    /**
     * 仅接受 [TaskStatus.PENDING]（重排队）与 [TaskStatus.CANCELLED]（取消），其余返回 `TASK:STATUS_CONFLICT`
     */
    readonly patchTask: (options: TaskControllerOptions['patchTask']) => Promise<
        AsyncTaskDto['TaskController/DEFAULT_TASK_FETCHER']
    > = async(options) => {
        let _uri = '/api/tasks/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'PATCH', body: options.body})) as Promise<AsyncTaskDto['TaskController/DEFAULT_TASK_FETCHER']>;
    }
    
    readonly transitionTaskStatuses: (options: TaskControllerOptions['transitionTaskStatuses']) => Promise<
        TaskStatusTransitionResponse
    > = async(options) => {
        let _uri = '/api/tasks/status-transitions';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<TaskStatusTransitionResponse>;
    }
}

export type TaskControllerOptions = {
    'createTask': {
        readonly body: TaskCreateRequest
    }, 
    'listTasks': {
        readonly parentId?: number | undefined, 
        readonly rootsOnly?: boolean | undefined, 
        readonly namespace?: string | undefined, 
        readonly taskType?: string | undefined, 
        readonly statuses?: ReadonlyArray<TaskStatus> | undefined, 
        readonly pageIndex?: number | undefined, 
        readonly pageSize?: number | undefined
    }, 
    'getTask': {
        readonly id: number
    }, 
    'getTaskTree': {
        readonly id: number
    }, 
    'transitionTaskStatuses': {
        readonly body: TaskStatusTransitionRequest
    }, 
    'patchTask': {
        readonly id: number, 
        readonly body: TaskStatusPatchRequest
    }
}
